### PR TITLE
Make ATen HIPify out-of-place, but still reuse CUDA names.

### PR DESCRIPTION
**Pull request by ezyang**
```
    This diff changes the HIPification of ATen to be out-of-place.
    We now have the following mappings:
    
    - ATen/cuda => ATen/hip
    - ATen/native/cuda => ATen/native/hip
    - ATen/native/sparse/cuda => ATen/native/sparse/hip
    - THC => THH
    - THCUNN => THHUNN
    
    The build system is adjusted to know about these new build paths,
    and HIPify is taught how to adjust include paths and
    THC_GENERIC_FILE appropriately.  ATen_hip is now built as
    the ATen_hip library, rather than reusing ATen_cuda.
    
    However, despite these new filepaths, none of the identifiers in ATen
    have actually changed.  So, e.g., THHGeneral.h still defines functions
    named THC_blahblah, and HIP still shows up as CUDA in PyTorch itself.
    We'll tackle this in a subsequent PR; this diff is just to get the files
    out-of-place.
    
    Minor extra improvements:
    
    - Don't edit tmp_install when hipifying
    - HIP no longer builds native_cudnn_cpp; it was unnecessary
    - Caffe2_HIP_INCLUDES is now Caffe2_HIP_INCLUDE, for consistency
      with all the other variables.
    - HIP build now properly respects ATEN_CUDA_FILES_GEN_LIB (it
      did not previously.)
    - You can now override file extension matching in pyHIPIFY
      by explicitly specifying its full name in the matching list.
      This is used so we can HIPify CMakeLists.txt in some situations.
    
    A little bit of string and ceiling wax:
    
    - gen.py grows a --rocm flag so that it knows to generate CUDA
      files which actually refer to the HIP headers (e.g., THH.h)
      We'll get rid of this eventually and generate real HIP files,
      but not for this PR.
    - Management of HIP dependencies is now completely deleted
      from the ATen CMakeLists.txt.  The old code was dead (because
      it was shoveled in ATen_CUDA_DEPENDENCY_LIBS and promptly
      ignored by the Caffe2 build system) and didn't actually work.
```

Stacked on https://github.com/pytorch/pytorch/pull/14849 review last commit only

### CODE REVIEW

File: aten/src/ATen/CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,9 @@ ENDIF()
 
 CONFIGURE_FILE(Config.h.in "${CMAKE_CURRENT_SOURCE_DIR}/Config.h")
 CONFIGURE_FILE(cuda/CUDAConfig.h.in "${CMAKE_CURRENT_SOURCE_DIR}/cuda/CUDAConfig.h")
+if(USE_ROCM)
```
Comment (bddppq) [inline aten/src/ATen/CMakeLists.txt]: Do we still need the CUDAConfig.h when USE_ROCM=ON?

File: aten/src/ATen/CMakeLists.txt [hunk 2]
```diff
@@ -28,21 +31,32 @@ FILE(GLOB cuda_cpp "cuda/*.cpp" "cuda/detail/*.cpp")
...
+FILE(GLOB native_cuda_cpp "native/cuda/*.cpp")
+FILE(GLOB native_cudnn_cpp "native/cudnn/*.cpp")
 FILE(GLOB native_sparse_cuda_cu "native/sparse/cuda/*.cu")
 FILE(GLOB native_sparse_cuda_cpp "native/sparse/cuda/*.cpp")
-FILE(GLOB native_cudnn_cpp "native/cudnn/*.cpp")
+
+FILE(GLOB native_hip_hip "native/hip/*.hip")
+FILE(GLOB native_hip_cpp "native/hip/*.cpp")
+FILE(GLOB native_sparse_hip_hip "native/sparse/hip/*.hip")
+FILE(GLOB native_sparse_hip_cpp "native/sparse/hip/*.cpp")
```
Comment (bddppq) [inline aten/src/ATen/CMakeLists.txt]: nit reorder these lines to match the cuda equivalent (e.g. currently cudnn is the third one in the above section while miopen is the last one in this hip section)

**Review by bddppq — APPROVED** [review]
Everything except gen.py looks good to me.

File: aten/src/ATen/CMakeLists.txt (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/CMakeLists.txt]: I think there might be some CPU files that rely on it. I'll check in another PR.

Comment (ezyang) [reply aten/src/ATen/CMakeLists.txt]: meh, rebasing, I'll check

File: aten/src/ATen/CMakeLists.txt [hunk 3]
```diff
@@ -81,21 +107,28 @@ IF(BLAS_FOUND)
     MESSAGE(STATUS "TH_BINARY_BUILD detected. Enabling special linkage.")
     list(APPEND ATen_CPU_DEPENDENCY_LIBS
       "${BLAS_LIBRARIES};${BLAS_LIBRARIES};${BLAS_LIBRARIES}")
-    if(USE_CUDA OR USE_ROCM)
+    if(USE_CUDA)
       list(APPEND ATen_CUDA_DEPENDENCY_LIBS
         "${BLAS_LIBRARIES};${BLAS_LIBRARIES};${BLAS_LIBRARIES}")
     endif()
+    if(USE_HIP)
```
Comment (bddppq) [inline aten/src/ATen/CMakeLists.txt]: USE_ROCM

File: aten/src/ATen/CMakeLists.txt (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/CMakeLists.txt]: Yeah, we do need it. I added a comment about it. We can get rid of it when out-of-place hipify happens.

**Pull request closed without merging**